crates/environ/src/module.rs

```diff
@@ -83,6 +79,164 @@ impl MemoryPlan {
     }
 }
 
+/// A WebAssembly linear memory initializer.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MemoryInitializer {
+    /// The index of a linear memory to initialize.
+    pub memory_index: MemoryIndex,
+    /// Optionally, a global variable giving a base index.
+    pub base: Option<GlobalIndex>,
+    /// The offset to add to the base.
+    pub offset: usize,
+    /// The data to write into the linear memory.
+    pub data: Box<[u8]>,
+}
+
+impl From<DataInitializer<'_>> for MemoryInitializer {
+    fn from(initializer: DataInitializer) -> Self {
+        Self {
+            memory_index: initializer.memory_index,
+            base: initializer.base,
+            offset: initializer.offset,
+            data: initializer.data.into(),
+        }
+    }
+}
+
+/// The type of WebAssembly linear memory initialization.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum MemoryInitialization {
+    /// Memory initialization is paged.
+    ///
+    /// To be paged, the following requirements must be met:
+    ///
+    /// * All data segments must reference defined memories.
+    /// * All data segments must not use a global base.
+    /// * All data segments must be in bounds.
+    ///
+    /// Paged initialization is performed by memcopying individual pages to the linear memory.
+    Paged {
+        /// The size of each page stored in the map.
+        /// This is expected to be the host page size.
+        page_size: usize,
+        /// The map of defined memory index to a list of page data.
+        /// The list of page data is sparse, with None representing a zero page.
+        /// The size of the list will be the maximum page written to by a data segment.
+        map: PrimaryMap<DefinedMemoryIndex, Vec<Option<Box<[u8]>>>>,
```

Comment:
I'll update the comment on why "paged" initialization is necessary, which is primarily for the uffd feature so that no work has to be performed in the user fault handler to initialize the page (we just tell the kernel "here's a page you can copy to the page being accessed").